src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -707,120 +925,94 @@ namespace nbl::ext::imgui
-					{
-						// Negative offsets are illegal for vkCmdSetScissor
-						if (clip_rect.x < 0.0f)
-							clip_rect.x = 0.0f;
-						if (clip_rect.y < 0.0f)
-							clip_rect.y = 0.0f;
-
-						{// Apply scissor/clipping rectangle
-							VkRect2D scissor{};
-							scissor.offset.x = static_cast<int32_t>(clip_rect.x);
-							scissor.offset.y = static_cast<int32_t>(clip_rect.y);
-							scissor.extent.width = static_cast<uint32_t>(clip_rect.z - clip_rect.x);
-							scissor.extent.height = static_cast<uint32_t>(clip_rect.w - clip_rect.y);
-							commandBuffer->setScissor(0, 1, &scissor);
-						}
-
-						// Draw
-						commandBuffer->drawIndexed(
-							pcmd->ElemCount,
-							1,
-							pcmd->IdxOffset + global_idx_offset,
-							pcmd->VtxOffset + global_vtx_offset,
-							0
-						);
-					}
-				}
-				global_idx_offset += cmd_list->IdxBuffer.Size;
-				global_vtx_offset += cmd_list->VtxBuffer.Size;
+			scratchSemaphoreCounters.newie = info.scratchSemaphore.value;
+
+			if (scratchSemaphoreCounters.newie != scratchSemaphoreCounters.oldie)
+			{
+				const auto overflows = scratchSemaphoreCounters.newie - scratchSemaphoreCounters.oldie;
+				logger->log("%d overflows when rendering UI!\n", nbl::system::ILogger::ELL_PERFORMANCE, overflows);
+
+				// TODO: handle them?
+				exit(0x45);
 			}
-		}
 
+			auto waitInfo = info.getFutureScratchSemaphore();
+			m_mdi.streamingTDBufferST->multi_deallocate(MDI_ALLOCATION_COUNT, multiAllocParams.offsets.data(), multiAllocParams.byteSizes.data(), waitInfo);
+		}
+	
 		return true;
-
 	}
 
-	void UI::Update(float const deltaTimeInSec, float const mousePosX, float const mousePosY, size_t const mouseEventsCount, ui::SMouseEvent const * mouseEvents) // TODO: Keyboard events
+	bool UI::update(const ui::IWindow* window, float const deltaTimeInSec, const core::SRange<const nbl::ui::SMouseEvent> mouseEvents, const core::SRange<const nbl::ui::SKeyboardEvent> keyboardEvents)
 	{
+		if (!window)
+			return false;
+
 		auto & io = ImGui::GetIO();
+
 		io.DeltaTime = deltaTimeInSec;
-		io.DisplaySize = ImVec2(m_window->getWidth(), m_window->getHeight());
+		io.DisplaySize = ImVec2(window->getWidth(), window->getHeight());
 
-		HandleMouseEvents(mousePosX, mousePosY, mouseEventsCount, mouseEvents);
+		handleMouseEvents(mouseEvents, window);
+		handleKeyEvents(keyboardEvents);
 
 		ImGui::NewFrame();
-		hasFocus = false;
+
 		for (auto const& subscriber : m_subscribers)
 			subscriber.listener();
 
-		ImGui::Render(); // note it doesn't touch GPU or graphics API at all, internal call - open the function def for more details
-	}
-
-	void UI::BeginWindow(char const* windowName)
-	{
-		ImGui::Begin(windowName);
-	}
-
-	void UI::EndWindow()
-	{
-		if (ImGui::IsWindowFocused())
-			hasFocus = true;
+		ImGui::Render(); // note it doesn't touch GPU or graphics API at all, internal call for IMGUI cpu geometry buffers update
```

Comment:
I'd have this sit in our `render` (maybe we can skip a memcpy)

Reply:
resolved

> (maybe we can skip a memcpy)

this part need to think of